### PR TITLE
Feature/5 3/background tasks interfaces

### PR DESCRIPTION
**Pull request by chfsx**
As discussed in the JF (9.1.2017) we move the technical discussion of the deployment of BackgroundTasks as a general Service to github since we're able to discuss more precise in the code.

there are currently several questions in http://www.ilias.de/docu/goto_docu_wiki_wpage_4384_1357.html
I will respond to them here later this week

An Example using the new interfaces and namespaceing is available at: https://github.com/studer-raimann/ILIAS/blob/feature/5-3/BackgroundTasksInterfacesAndBasicImplementation/src/BackgroundTasks/Examples/SimpleExample.php

### CODE REVIEW

File: src/BackgroundTasks/Task/UserInteraction.php [hunk 1]
```diff
@@ -22,8 +22,8 @@ public function getOptions();
 
 	/**
 	 * @param \ILIAS\BackgroundTasks\IO $input
-	 * @param string $user_seleced_option
-	 * @return \ILIAS\BackgroundTasks\IO
+	 * @param $user_selected_option
```
Comment (chfsx) [inline src/BackgroundTasks/Task/UserInteraction.php]: still a string, isn't it?

File: src/BackgroundTasks/Worker.php [hunk 2]
```diff
@@ -13,4 +13,11 @@
 	 * @return void
 	 */
 	public function doWork();
+
+	/**
+	 * Returns true iff the worker wants to be called within the current HTTP request.
```
Comment (chfsx) [inline src/BackgroundTasks/Worker.php]: typo iff --> if

File: src/BackgroundTasks/BackgroundTaskManager.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ */
+interface BackgroundTaskManager {
+
+	/**
+	 * @param $bucket Bucket
+	 * @param $user_ids int[]
+	 * @return Bucket
+	 */
+	public function putInQueueAndObserve($bucket, $user_ids);
```
Comment (chfsx) [inline src/BackgroundTasks/BackgroundTaskManager.php]: this should throw an exception when no observer is registred, I think

Comment (chfsx) [reply src/BackgroundTasks/BackgroundTaskManager.php]: typehinting on Bucket?

File: src/BackgroundTasks/Task.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	/**
+	 * @return bool
+	 */
+	public function isUserInteraction();
+
+
+	/**
+	 * @return bool
+	 */
+	public function isJob();
```
Comment (alex40724) [inline src/BackgroundTasks/Task.php]: Are these isUserInteraction() and isJob() functions needed? If an object implements Task/Job it's a Job and if it's implementing Task/Userinteraction it's a UserInteraction. So using instanceof should be used if this information is needed.

File: src/BackgroundTasks/Worker.php [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ */
+interface Worker {
+
+	/**
+	 * @return void
+	 */
+	public function doWork();
+
+	/**
+	 * Returns true iff the worker wants to be called within the current HTTP request.
```
Comment (klees) [inline src/BackgroundTasks/Worker.php]: Since Oskar wrotes this, I assume that `iff` means `if and only if`? Is this something every ILIAS dev will recognize?

File: src/BackgroundTasks/Exception.php [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+require_once('Services/Exceptions/classes/class.ilException.php');
+
+/**
+ * Class Exception
+ *
+ * @package ILIAS\BackgroundTasks
+ *
+ *          The Basic Exception Class for BackgroundTasks. PLease Specify by extending
+ */
+class Exception extends \ilException {
```
Comment (klees) [inline src/BackgroundTasks/Exception.php]: + `abstract`, I think.

Comment (klees) [reply src/BackgroundTasks/Exception.php]: Should we somehow specify, which Exceptions should derive from this? To me it looks as these should only be Exceptions in the general handling of background tasks, not exceptions in specific actions like ZipFolder, right?

File: src/BackgroundTasks/Exception.php [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+	const E_BASIC = 10001;
```
Comment (klees) [inline src/BackgroundTasks/Exception.php]: What is this good for?

File: src/BackgroundTasks/Task.php (hunk 4 above)
Comment (klees) [reply src/BackgroundTasks/Task.php]: I somehow second this, but would also go further: Why would you ever want to now this (may it by using isJob() or by using instanceof)? In most cases it is imho better to instead define the behaviour you want to switch on the class itself. Fabian and I had a similar discussion for the table lock implementation, he might remember.

File: src/BackgroundTasks/Task/UserInteraction.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+	 */
+	public function getOptions();
+
+
+	/**
+	 * @param \ILIAS\BackgroundTasks\IO $input
+	 * @param $user_selected_option
+	 * @return IO
+	 */
+	public function interaction(IO $input, $user_selected_option);
```
Comment (klees) [inline src/BackgroundTasks/Task/UserInteraction.php]: What does this do?

File: src/BackgroundTasks/Worker.php [hunk 9]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace ILIAS\BackgroundTasks;
+
+/**
+ * Interface Worker
+ *
+ * @package ILIAS\BackgroundTasks
+ */
+interface Worker {
```
Comment (klees) [inline src/BackgroundTasks/Worker.php]: What is this?

File: src/BackgroundTasks/Bucket.php [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+ *
+ * @package ILIAS\BackgroundTasks
+ *
+ *          A Bucket is used to pack several Tasks which have to be done one after the other and is put in the background to avoid hight impact on
+ *          the ILIAS system.
+ *
+ *          Whenever packing a bucket, you start with an Input (IO-interface) wich will be passed to the first task. your task will provide an output
+ *          (IO-interface again) wich will be passed to the next Task, ...
+ *
+ *          Bucket (IO -> Task -> IO -> Task)
```
Comment (klees) [inline src/BackgroundTasks/Bucket.php]: Question to Oskar, to be discussed while having a beer, maybe: Did you notice, that the Tasks most probably should have a monadic structure?

I feel this "pipe" pattern won't be enough really soon. It also makes it near impossible for the engine part of the background task to know anything about the innards of the task and also spoils composability and reusability on the other side. I.e., if I want to switch the next task based on the task before, I would need to create a new task class and make the switch somewhere inline. I would then need to instantiate the other task by myself and somehow run it. I don't see how this would work currently.

File: src/BackgroundTasks/BackgroundTaskManager.php (hunk 3 above)
Comment (otruffer) [reply src/BackgroundTasks/BackgroundTaskManager.php]: Yes, if the user_ids cannot be resolved to a user we need to throw an exception!

File: src/BackgroundTasks/Worker.php (hunk 2 above)
Comment (otruffer) [reply src/BackgroundTasks/Worker.php]: "if" stands for "if": =>
"iff" stands for "if and only if": <=>

File: src/BackgroundTasks/Task/UserInteraction.php (hunk 1 above)
Comment (otruffer) [reply src/BackgroundTasks/Task/UserInteraction.php]: Yes, will be altered to a Option Container

File: src/BackgroundTasks/Task.php (hunk 4 above)
Comment (otruffer) [reply src/BackgroundTasks/Task.php]: You're right.

Back in university somewhere around year 1 I was told: Avoid instanceofs! And somehow that still clings to me even if "insanceof" makes more sense :-)

File: src/BackgroundTasks/Worker.php (hunk 2 above)
Comment (chfsx) [reply src/BackgroundTasks/Worker.php]: thanks, you never stop learning ;)

File: src/BackgroundTasks/Observer.php [hunk 11]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace ILIAS\BackgroundTasks;
+
+/**
+ * Interface Observer
+ *
+ * @package ILIAS\BackgroundTasks
+ */
+interface Observer {
```
Comment (klees) [inline src/BackgroundTasks/Observer.php]: What is this? By reading the Code I came to the conclusion that this basically says who could fill out the required user inputs and see the progress and result of the task. Is this correct? Please mention this in the docstring of the class.

File: src/BackgroundTasks/Worker.php [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+	 * @return void
+	 */
+	public function doWork();
+
+	/**
+	 * Returns true iff the worker wants to be called within the current HTTP request.
+	 *
+	 * @return boolean
+	 */
+	public function isSynchronised();
```
Comment (klees) [inline src/BackgroundTasks/Worker.php]: What does synchronised mean here?

File: src/BackgroundTasks/IO.php [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * Interface IO
+ *
+ * @package ILIAS\BackgroundTasks
+ *
+ *          The IO as a defined format of data passed between two tasks. IO MUST be serialisable since it will bes stored in the database or
+ *          somewhere else
+ */
+interface IO extends \Serializable {
```
Comment (klees) [inline src/BackgroundTasks/IO.php]: To me this looks as if it should be named `Value`. I also feel that we will need this kind of things in other places soon (async, forms), so we might want to put this in a separate service. I think @Amstutz is playing with a similar thing.

File: src/BackgroundTasks/IO.php [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+
+	/**
+	 * @return string Gets a hash for this IO. If two objects are the same the hash must be the same! if two objects are different you need to have
+	 *                as view collitions as possible.
+	 */
+	public function getHash();
```
Comment (klees) [inline src/BackgroundTasks/IO.php]: I understand the "Serializable", but what is this good for? Why must this interface implement equals if you already have the hash that you could compare?

File: src/BackgroundTasks/IO.php [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+	 * @param \ILIAS\BackgroundTasks\IO $other
+	 * @return mixed
+	 */
+	public function equals(IO $other);
+
+
+	/**
+	 * @var string get the Type of the
+	 */
+	public function getType();
```
Comment (klees) [inline src/BackgroundTasks/IO.php]: What is this? Why could it be different from the class name?

File: src/BackgroundTasks/Task/Job.php [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+	 * @param \ILIAS\BackgroundTasks\IO $input
+	 * @return \ILIAS\BackgroundTasks\IO
+	 */
+	public function run(IO $input);
+
+
+	/**
+	 * @return bool Returns true iff the job supports giving feedback about the percentage done.
+	 */
+	public function supportsPercentage();
```
Comment (klees) [inline src/BackgroundTasks/Task/Job.php]: A job not supporting percentage could still be "done" or "not done", right? We could map this to "0%" and "100%" and get rid of this then.

File: src/BackgroundTasks/Task/Job.php [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+
+	/**
+	 * @return int Returns 0 if !supportsPercentage and the percentage otherwise.
+	 */
+	public function getPercentage();
```
Comment (klees) [inline src/BackgroundTasks/Task/Job.php]: Could we move this to Task?

Comment (klees) [reply src/BackgroundTasks/Task/Job.php]: How would i use this and/or implement this? I'm currently wondering, about this: I have a php process, that somehow instantiated the Job and called `$job->run`. Where would i use `$job->getPercentage` then? Other way round: If I would want to use this from another process: How would the two instances of the job be identified? Where, as the job, would I put my percentage info? Would everyone need to implement this on its own?

Comment (klees) [reply src/BackgroundTasks/Task/Job.php]: Did you think about a model akin to the Cronjob, where I call a ping method from within the job? Some service supporting this should then be passed to `run` imo.

**Review by klees — CHANGES_REQUESTED** [review]
Thanks for the PR with interfaces only, that helped a lot. I see a lot of open questions regarding the proposed interface and also have some general questions and remarks that I will put in a separate comment, as they are not really tied to a location in the code.

File: src/BackgroundTasks/Worker.php (hunk 9 above)
Comment (otruffer) [reply src/BackgroundTasks/Worker.php]: Added a description

File: src/BackgroundTasks/Worker.php (hunk 5 above)
Comment (otruffer) [reply src/BackgroundTasks/Worker.php]: @chfsx Did not know about this, I thought it was a well known abbreviation. I'd prefer if we could use it as it's extremly comfy to use. Iff that won't be the case i'll be sad.

File: src/BackgroundTasks/Worker.php (hunk 12 above)
Comment (otruffer) [reply src/BackgroundTasks/Worker.php]: Would you prefer a different wording or a different php doc?

File: src/BackgroundTasks/Bucket.php (hunk 10 above)
Comment (otruffer) [reply src/BackgroundTasks/Bucket.php]: Yes it suddenly jumps into your face after a while. I think that's why i subconciously named the value containers IO. What we would really want is pretty much the IO Monad from haskell together with its "do" syntax sugar. And it should be serializable. 

The pipe was enough for unix ;-) But I agree, it's just really hard to think of an interface that is easy to use and complete enough for all possible. Maybe:
$bucket = new Bucket();
$add = new AddJob();
$a = new IntegerValue(1);
$b = new IntegerValue(2);
$bucket->addTask("variable1", $add, [$a, $b]);
$bucket->addTask("returnValue", $add, [$a, "variable1"]);
$bucket->setReturnValue("returnValue");
with 
```
/**
* @param $variableName string
* @param $task Task
* @param $parameters (string|Value)[] If string => variable if Value => constant.
**/
addTask($variableName, $task, $parameters);
```

File: src/BackgroundTasks/Observer.php (hunk 11 above)
Comment (otruffer) [reply src/BackgroundTasks/Observer.php]: Done.

File: src/BackgroundTasks/IO.php (hunk 13 above)
Comment (otruffer) [reply src/BackgroundTasks/IO.php]: Yeah, I don't really care about the name. Value sounds indeed better. @Amstutz I'm interested in your approach! :+1:

File: src/BackgroundTasks/IO.php (hunk 15 above)
Comment (otruffer) [reply src/BackgroundTasks/IO.php]: Probably not, I can't remember why I put it there. May have had my reasons but they are only in past me's head.

File: src/BackgroundTasks/Task/Job.php (hunk 16 above)
Comment (otruffer) [reply src/BackgroundTasks/Task/Job.php]: Concerning the UI it will look nicer if you can display a percentage independent loading bar instead of just jumping from 0 to 100.

File: src/BackgroundTasks/Task/Job.php (hunk 17 above)
Comment (otruffer) [reply src/BackgroundTasks/Task/Job.php]: That's a good idea! I'll pass the observer to the run method, and add a ping method to the observer.

**Comment by klees** [comment]
Hi Fabian, hi Oskar,

I promised some general remarks and questions, so here we go:

* Did you look into other libraries/apps solving similar problems? I see that you are indeed using some terminology (Worker, Task), so I think you did. But which? Could you comment on how they compare to your proposal?
* I think this will benefit from a separation into runtime (worker, manager, ...) and "task construction". I would also suggest to factor out this "IO" thingy, as I already mentioned in the comments.
* After reading the entry in the feature wiki, I suggested to let Bucket inherit from Task (which you did, thx). Now I see, that this was a little misguided. I think the bucket somehow mixes two concerns that could be separated for good. The one concern is, that one or more tasks should be combinable to an overall task, that was what I had seen in the FR. The other concern seems to be, that the bucket captures one execution of a task. I didn't notice that before.
* As already suggested in some mails, the task-part of the background tasks should be abstractable in a way that merges with the activities in the workflow engine.

Best regards!

**Comment by otruffer** [comment]
> Did you look into other libraries/apps solving similar problems? I see that you are indeed using some terminology (Worker, Task), so I think you did. But which? Could you comment on how they compare to your proposal?

I did not find an implementation in a framework that matches our exact needs: Composable tasks that can be laid into the background.

> I think this will benefit from a separation into runtime (worker, manager, ...) and "task construction". I would also suggest to factor out this "IO" thingy, as I already mentioned in the comments.

Agreed, We'll reword it to "Value"

> After reading the entry in the feature wiki, I suggested to let Bucket inherit from Task (which you did, thx). Now I see, that this was a little misguided. I think the bucket somehow mixes two concerns that could be separated for good. The one concern is, that one or more tasks should be combinable to an overall task, that was what I had seen in the FR. The other concern seems to be, that the bucket captures one execution of a task. I didn't notice that before.

Good idea, we should conceptually separate execution (run-time) from construction (compile-time) by namespacing/foldering the interfaces further. When it comes to the buckets, I think it will not matter: When constructing the background task you might want to reuse other buckets created by a method and combine buckets and tasks as you go. When serializing the top-level-bucket (containing other buckets and tasks recursivly) into the queue we might want to flatten the data structure into one bucket containing tasks. Thus in the "run-time" we get one bucket with a list of tasks. We still want that top level bucket as it will be similar to a closure: When we want to delete an operation (e.g. something went wrong) we want to identify all operations and values related to the bucket and wipe them.

> As already suggested in some mails, the task-part of the background tasks should be abstractable in a way that merges with the activities in the workflow engine.

Add the moment i'd suggest to add the Activity interface to the Job Interface. This way we should be able to reuse Jobs as Workflow Engine activities.

File: src/BackgroundTasks/Task/UserInteraction.php [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+use ILIAS\BackgroundTasks\Observer;
+use ILIAS\BackgroundTasks\Task;
+
+/**
+ * Interface UserInteraction
+ *
+ * @package ILIAS\BackgroundTasks\Task
+ *
+ *          A Task in the Bucket, which will need some User-Interaction before running the task. A
+ *          User-Interaction is provided as
```
Comment (chfsx) [inline src/BackgroundTasks/Task/UserInteraction.php]: @otruffer you just stopped writing within the sentence ;-)

**Comment by otruffer** [comment]
Some information before reviewing the Code:

Currently there are implementations for task definition and composition. The tasks cannot be stored in the database yet. There is no implementation of a bucket yet. The tasks can be executed with the simplistic BasicTaskManager, this is just for testing currelty. 

For restoring creating tasks we use dependency injection (newly implemented in this branch) see /src/DI/factory.php. This makes testing possible for this module.

For testing Jobs we implemented a PlusJob  (see src/BackgroundTasks/Implementation/Tasks/PlusJob.php), which takes two integers and returns one integer. In the plus job you can easily see how you will implement tasks.

For composition testing we implemented a few tests (see tests/BackgroundTask). Most interesting here will be the TaskTest for the moment as it shows Task composition. The simplest task can be found in testValueWrapper()

```
                // The following part is only needed in tests. In production you will use global $factory.
                $dic = new Container();
		$dic[Observer::class] = function ($c) {
			return new ObserverMock();
		};
		$factory = new Factory($dic);

                // Here's the creation of the plus job and the setting of the values
		$t = $factory->createInstance(PlusJob::class);
		$t->setInput([1, 4]);
                
                // The following part is only needed in tests. In production you will use $taskManager->startTask($t), which will create a Bucket with one Task and enqueue the task.
		$taskManager = new BasicTaskManager();
		/** @var IntegerValue $finalValue */
		$finalValue = $taskManager->executeTask($t, new ObserverMock());
		$this->assertEquals($finalValue->getValue(), 5);
```

More complicated cases like putting tasks together can be found in the tests as well.

@ILIAS-eLearning/technical-board Would it be possible to discuss the job creation and composition at the technical board?

**Review by Amstutz — CHANGES_REQUESTED** [review]
Please add a short readme with a short description and examples making the usage of this clear.

Note that we state in a rule for the ./src folder (see the according [README.md](https://github.com/ILIAS-eLearning/ILIAS/blob/release_5-2/src/README.md) there) we state that:
`6. Libraries [in the src dirctory] SHOULD contain a README.md on the top-level, that explains the purpose and design of the library and gives usage examples.`

It is not absolutely mandatory, but in this rather complex case, I would heavily recommend to do so.

**Comment by otruffer** [comment]
Okay: Brough everything up to date, added a readme and rebased. Now this pull request only contains the interfaces.

**Comment by Amstutz** [comment]
Thank you very much for the readme!

**Review by Amstutz — APPROVED** [review]

**Comment by chfsx** [comment]
Hi @klees and @alex40724 
We would like to proceed with this PR and merge it after at least 2 core-developers approved (with 2/3 over-all approvals).

Thanks you very much for your effort on reviewing this, I know this always takes some time and therefore I hope you could again have a look at it. Best regards!

**Comment by chfsx** [comment]
Since i opened the PR I'm not allowed on github to approve it, so I do it here: I approve this PR! thanks you very much @otruffer

**Comment by alex40724** [comment]
The readme refers to an example implementation PlusJob.php - but this is not included, is it? Or has it been removed since the PR "should only include the interfaces"? I put a lot of implementation into my PRs for the KS. I thought this would help to understand the details sometime. For me every line of code helps...

I currently do not object to anything here. But I would prefer that @klees aproves this, too. Since he discussed the details more than I did.

**Comment by chfsx** [comment]
no, it's only a fictional example to show how
it works. we provide only the interfaces in this PR (and the readme)

**Comment by alex40724** [comment]
Yes, I just saw the earlier comment (and edited my previous one). Thanks to all that put effort into this non trivial stuff. I approve, but suggest to wait for @klees comments...

**Comment by klees** [comment]
I'm on it.

File: src/BackgroundTasks/BucketMeta.php [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+use ILIAS\BackgroundTasks\Task\UserInteraction\Option;
+
+/**
+ * Interface Observer
+ *
+ * @package ILIAS\BackgroundTasks
+ *
+ * A meta bucket contains infos about a bucket like its percentage, name etc.
+ */
+interface BucketMeta {
```
Comment (klees) [inline src/BackgroundTasks/BucketMeta.php:15]: As far as I understand, this duplicates some methods from the bucket. Let `Bucket` return an instance of this or extend `Bucket` from this. Maybe also rename this to `BucketInfo`.

File: src/BackgroundTasks/Persistence.php [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+ * Created by PhpStorm.
+ * User: otruffer
+ * Date: 05.05.17
+ * Time: 10:14
+ */
+namespace ILIAS\BackgroundTasks;
+
+use ILIAS\BackgroundTasks\Implementation\Bucket\State;
+
+interface Persistence {
```
Comment (klees) [inline src/BackgroundTasks/Persistence.php:12]: Yeah, I like this pattern. For future reference: I think this is the "Repository Pattern". It helps testability a lot, imo, when contrasted with "Active Record".

File: src/BackgroundTasks/Task.php [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+ *          - Job: A Task, which can be run without any interaction with the user such as zipping
+ *          files or just collecting some data
+ *          - UserInteraction: A Task in the Bucket, which will need some User-Interaction before
+ *          running the task. A User-Interaction is provided as Button in the UserInterface such as
+ *          [ Cancel ] or [ Download ]
+ */
+interface Task {
+
+	/**
+	 * @return string
```
Comment (klees) [inline src/BackgroundTasks/Task.php:23]: This should be `Type`, right?

Comment (klees) [reply src/BackgroundTasks/Task.php:23]: Ok, I guess it should not be `Type`, because we also have `getOutputType`. What is this, then? Is this the class of the `Task`? Why would I want to use this, then, instead of `get_class`?

File: src/BackgroundTasks/Task/Job.php [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+use ILIAS\BackgroundTasks\Value;
+use ILIAS\BackgroundTasks\Task;
+
+/**
+ * Interface Job
+ *
+ * @package ILIAS\BackgroundTasks\Task
+ *
+ *          A Task, which can be run without any interaction with the user, such as zipping files
+ *          or just collecting some data
```
Comment (klees) [inline src/BackgroundTasks/Task/Job.php:15]: Nice explanation. Maybe move this to your glossary as well, the "without interaction" part is missing there.

File: src/DI/Factory.php [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
+ * Class Factory
+ * @package ILIAS\DI
+ *
+ * Create instances of classes using type hinting and the dependency injection container.
+ *
+ * @author Oskar Truffer <ot@studer-raimann.ch>
+ */
+class Factory {
```
Comment (klees) [inline src/DI/Factory.php]: I do understand that you need this kind of factory to initialize the tasks (there are other options, still), but I'm against considering this a general pattern for instantiation. Thus move this to the Background-Tasks-folder, please. The arguments I see against this kind of factory are:
* Folklore says, ReflectionClass is slow. This is not too relevant for the background tasks, but might be a problem when this factory is used in ordinary requests. We should measure the impact of ReflectionClass before  we make this generally available.
* More severe: This pattern confuses "implements XYZ" with "is named XYZ". This is not the same. By announcing a dependency in a constructor, one (optimally) uses interfaces to describe the services an object depends on. This is the "implement XYZ" part. The array-access on the DIC, on the contrary, gives names to certain services. These are optimally only temporary/local when DI is used properly, but ILIAS is not there yet. The problem with this confusion is: the interfaces a service implements only accidentally coincide with its name. There already are services in ILIAS where this is in fact wrong (`$tpl` vs. `ilTemplate`). When announcing a dependency, I should only say: "I need some service implementing interface XYZ", but with this pattern I, in fact, say "I need the service named XYZ". This will prevent some moves in the future, e.g. having two services implementing the same interface.

**Review by klees — CHANGES_REQUESTED** [review]
Hi @otruffer,
some nitpicks and suggestions, one major problem (DI/Factory). Thanks for the good work!
Regards!

File: src/BackgroundTasks/Persistence.php (hunk 20 above)
Comment (otruffer) [reply src/BackgroundTasks/Persistence.php:12]: Yup, made the whole testing possible. Repository might have been a better name then :-)

File: src/BackgroundTasks/Task.php (hunk 21 above)
Comment (otruffer) [reply src/BackgroundTasks/Task.php:23]: Is currently "$StringRepresentationOfInputType -> $StringRepresentationOfOutputType". It's not really necessary but for exception text.

File: src/DI/Factory.php (hunk 23 above)
Comment (otruffer) [reply src/DI/Factory.php]: > Folklore says, ReflectionClass is slow. This is not too relevant for the background tasks, but might be a problem when this factory is used in ordinary requests. We should measure the impact of ReflectionClass before we make this generally available.

Injection needs reflection and the benchmarks I have found (admittedly they do not seem very thorough) indicate that the overhead generated by the reflection is neglectable. If we don't want reflection, we need to rename the "Dependency Injection" service to "Dependency Container".

> More severe: This pattern confuses "implements XYZ" with "is named XYZ". This is not the same. By announcing a dependency in a constructor, one (optimally) uses interfaces to describe the services an object depends on. This is the "implement XYZ" part. The array-access on the DIC, on the contrary, gives names to certain services. These are optimally only temporary/local when DI is used properly, but ILIAS is not there yet. The problem with this confusion is: the interfaces a service implements only accidentally coincide with its name. There already are services in ILIAS where this is in fact wrong ($tpl vs. ilTemplate). When announcing a dependency, I should only say: "I need some service implementing interface XYZ", but with this pattern I, in fact, say "I need the service named XYZ". This will prevent some moves in the future, e.g. having two services implementing the same interface.

Here we have the problem that I provided the interface with the implementation. The Interface provided in this PR stayed the same the implementation was changed in my implementation branch. This is a mistake on my part, sorry. The current implementation uses a dependency map to match interfaces to instantiations in the DIC regarding the class that requests the dependency. Thus it will be possible to have two implementations of an interface and hand them out depending on who requests the dependency (the redis + mysql usecase we discussed). 

This said I'm not in favor nor against keeping this interface in the DI and can move it to the BT as well. I thought it might be of interest to provide a basic injector to actually justify the services name as a first step. If you are still against keeping this and we get no other feedback by tomorrow I'll move it to the BT service.

Thanks a lot for the review! :-)

Edit: I added the dependency map's interface for you to see.

**Comment by otruffer** [comment]
@klees Okay, factory now reads injector and is moved to the bt folder. Anything else?

**Review by klees — APPROVED** [review]

**Comment by chfsx** [comment]
Thanks everybody for the comments, reviews and all the brainstorming!
